### PR TITLE
[ml] `TorchPredictor` interface

### PR DESCRIPTION
**Pull request by amogkam**
Adds interface for `TorchPredictor`.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    ) -> "TorchPredictor":
+        """Instantiate the predictor from a Checkpoint.
+
+        The checkpoint is expected to be a result of ``TorchTrainer``.
+
+        Args:
+            checkpoint: The checkpoint to load the model and
+                preprocessor from. It is expected to be from the result of a
+                ``TorchTrainer`` run.
+            model_definition: If checkpoints
```
Comment (maxpumperla) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:35]: I think this needs to be clearer in terms of what `model_definition` actually is. Both `model` and this one are `torch.nn.Module`s and I'm not sure users will understand what's in our checkpoint and what's not.

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        self,
+        data: DataBatchType,
+        column_names: Optional[List[str]] = None,
+        feature_columns: Optional[Union[List[str], List[List[str]]]] = None,
+        dtype: Optional[Union[torch.dtype, List[torch.dtype]]] = None,
+    ) -> DataBatchType:
+        """Run inference on data batch.
+
+        Args:
+            data: A batch of input data.
```
Comment (maxpumperla) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:52]: I think this is currently defined as union of pandas dfs and numpy arrays, correct? If so, maybe we should add this information here. I know there's a risk of diverging if `DataBatchType` changes, but I'm not sure this abstraction is currently very helpful to end users.

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            columns_names: If provided, the column names to set for the data
```
Comment (maxpumperla) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:53]: What does that mean, to "set" them? Also, how does this relate to the next arg?

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                batch.
+            feature_columns: The names of the columns in the dataframe to
```
Comment (maxpumperla) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:55]: Either call this one `feature_column_names` or the other one `columns`, for consistency.

Also it says "dataframe" here, but my understanding is that we might have e.g. numpy arrays, too. Slightly confusing to me.

Also, is it a conscious decision not to include our native `Dataset` abstraction in all of this?

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                use as features to predict on. If this arg is a List[List[
+                str]], then the data batch will be converted into a list of
+                tensors. This is useful for multi-input models. If None,
+                then use all columns in the ``data_batch``.
+            dtype: The torch dtype to use for the tensor. If set to None,
```
Comment (maxpumperla) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:60]: The return type is `DataBatchType` (numpy and pandas), so how does all of this relate to torch tensors and their types?

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class TorchPredictor(Predictor):
+    """A predictor for PyTorch models.
+
+    Args:
+        model: The torch module to use for predictions.
+        preprocessor: The preprocessor used to transform data batches prior
+            to prediction.
+
```
Comment (ericl) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:17]:
```suggestion
```

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                use as features to predict on. If this arg is a List[List[
+                str]], then the data batch will be converted into a list of
+                tensors. This is useful for multi-input models. If None,
+                then use all columns in the ``data_batch``.
+            dtype: The torch dtype to use for the tensor. If set to None,
+                then automatically infer the dtype.
+
+        Returns:
+            DataBatchType: Prediction result.
+
```
Comment (ericl) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:65]:
```suggestion
```

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import torch
+
+from ray.ml.predictor import Predictor, DataBatchType
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+
+
+class TorchPredictor(Predictor):
+    """A predictor for PyTorch models.
```
Comment (ericl) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:11]: Can you add an example of using TorchPredictor here with a simple model and data batch?

Comment (ericl) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:11]: (show both numpy and pandas inputs)

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+        """
+        raise NotImplementedError
+
+    def predict(
+        self,
+        data: DataBatchType,
+        column_names: Optional[List[str]] = None,
+        feature_columns: Optional[Union[List[str], List[List[str]]]] = None,
+        dtype: Optional[Union[torch.dtype, List[torch.dtype]]] = None,
```
Comment (simon-mo) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:47]: why can't these go into preprocessor? or the model itself (as first layer or in its call function)

Comment (simon-mo) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:47]: here we are overloading the predict interface, should we go back and add a **kwargs to the parent interface? (for example, in Serve's model wrapper, we expect Predictor.predict only take a single argument)

Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:47]: ah yeah **kwargs was already to Predictor.predict in a previous PR (https://github.com/ray-project/ray/commit/79a3b56015586a2a825ca52cc85184a0b47979a0)

Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:47]: @simon-mo can you raise the first point in the broader group for more visibility/discussion?

Comment (simon-mo) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:47]: ahhh got it. this resolved my second question

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+            to prediction.
+
+    """
+
+    def __init__(self, model: torch.nn.Module, preprocessor: Preprocessor):
+        raise NotImplementedError
+
+    @classmethod
+    def from_checkpoint(
+        cls, checkpoint: Checkpoint, model_definition: Optional[torch.nn.Module] = None
```
Comment (richardliaw) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:25]: will the model definition be a class or an object?

Comment (richardliaw) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:25]: - If it's a class, the typing should be Type[torch.nn.Module] right?

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 3 above)
Comment (richardliaw) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:53]: +1

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 4 above)
Comment (richardliaw) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:55]: +1 about first 2 points.

> Also, is it a conscious decision not to include our native Dataset abstraction in all of this?

Let's chat about this in a non-blocking way. 

My understanding is that we should look at datasets as a compute + data layer, and if you want to do large scale compute you should pass the model to dataset.map_batches. As such, we let predictors take single-threaded work, and the Dataset abstraction will handle scaling.

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 5 above)
Comment (richardliaw) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:60]: I'd opt for this also to be a non-blocking discussion.

I think the goal is for torch tensors to be converted into numpy/dataframe types before input / output across all predictors. As such, we can then build layers (like serving, etc) on top without doing too much special-casing.

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:52]: Yeah, we can add a short comment like "For example, a pandas df or numpy array"

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:35]: +1

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 9 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:47]: (Resolved offline)

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 10 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:25]: It would be an object- the actual object is needed to load the state dict, the class is not sufficient enough.

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 6 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py]: Fixed!

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 7 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py]: Fixed!

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 8 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:11]: Added example!

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 5 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py]: @maxpumperla I clarified this in the docstring, ptal!

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 4 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py]: `column_names` has been removed.

Changed the wording of dataframes.

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 3 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py]: `column_names` has been removed.

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 2 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py]: Clarified in the docstring!

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 1 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py]: Ok changed the name to `model`, does that help?

Also it would help if you guys could clarify on what's not clear. What's missing from the docstring in your mind?

Regarding whether users will understand what's in our checkpoint, they are the ones who are creating the checkpoint, so they should be aware of this.

**Review by richardliaw — APPROVED** [review]

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        data: DataBatchType,
+        feature_columns: Optional[
+            Union[List[str], List[List[str]], List[int], List[List[int]]]
+        ] = None,
+        dtype: Optional[torch.dtype] = None,
+    ) -> DataBatchType:
+        """Run inference on data batch.
+
+        The data is converted into a torch Tensor before being inputted to
+        the model.
```
Comment (Yard1) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:53]: So we are not supporting everything that's supported by `Dataset.to_torch`? I think it would make sense to reuse the logic and arguments.

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+        Args:
+            data: A batch of input data. Either a pandas DataFrame or numpy
+                array.
+            feature_columns: The names or indices of the columns in the
+                data to use as features to predict on. If this arg is a
```
Comment (maxpumperla) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:59]: what is this supposed to be for numpy arrays?

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+                list of lists, then the data batch will be converted into a
+                multiple tensors which are then concatenated before feeding
+                into the model. This is useful for multi-input models. If
+                None, then use all columns in ``data``.
+            dtype: The torch dtype to use when creating the torch tensor.
```
Comment (maxpumperla) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:64]: dtype of what? Input is `DataBatchType`, output too. we should say what mean be "the torch tensor". I think there can be multiple tensors, too, as per above "list of lists" definition. should we then also have lists of dtypes?

I fear this might go a little too far, but we need to anticipate questions like this.

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 12 above)
Comment (maxpumperla) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:59]: ah, so proper indices for numpy (int) and column names for pandas (str)? maybe clarify this even more!

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+            import torch
+            from ray.ml.predictors.torch import TorchPredictor
+
+            model = torch.nn.Linear(1, 1)
+            predictor = TorchPredictor(model=model)
+
+            # Pandas dataframe.
+            data = pd.DataFrame([[1, 2], [3, 4]], columns=["A", "B"])
+
+            predictions = predictor.predict(data)
```
Comment (maxpumperla) [inline python/ray/ml/predictors/integrations/torch/torch_predictor.py:96]: these examples help immensely!

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 12 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:59]: Yes! But column indices will also work for pandas dataframes if it doesn’t have any column names.

**Review by ericl — APPROVED** [review]

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 11 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:53]: The arguments that are not being included are either related to label columns (which is not relevant for inference) or specific to datasets (batch size, prefetch_blocks, etc.).

File: python/ray/ml/predictors/integrations/torch/torch_predictor.py (hunk 13 above)
Comment (amogkam) [reply python/ray/ml/predictors/integrations/torch/torch_predictor.py:64]: the multiple tensors are concatenated into a single tensor before being inputted to the model.
